module extension: add support for canonical name resolution

gabrielschulhof opened the pull request:
Before attempting to load a module, each provided resolver must be given an
opportunity to examine the name of the requested module without actually
loading it so as to canonicalize it, in case a module can be referred to by
multiple names.

Then, modules are loaded and cached by their canonical name.

LaszloLango commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -87,23 +104,39 @@ typedef jerry_value_t (*jerryx_native_module_on_resolve_t) (void);
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name);
+  bool (*resolve) (const jerry_value_t canonical_name,
+                   jerry_value_t *result);
+};
```
`} jerryx_module_resolver_t;`

gabrielschulhof commented:
@LaszloLango fixed.

zherczeg commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -130,20 +163,36 @@ load_and_evaluate_js_file (const jerry_char_t *name, jerry_value_t *result)
+{
+  jerry_value_t resulting_path;
+
+  /* Search name for "./" and "../", follow symlinks, etc., then create resulting_path via jerry_create_string () */
```
Full stop after sentence.

zherczeg commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -130,20 +163,36 @@ load_and_evaluate_js_file (const jerry_char_t *name, jerry_value_t *result)
 {
   /* Consult the JerryScript module resolver first, in case the requested module is a compiled-in JerryScript module. */
-  jerryx_module_native_resolver,
+  &jerryx_module_native_resolver,
```
Why do you need & ?

gabrielschulhof replied:
Because `jerryx_module_native_resolver` is now a structure stored in a global variable, and this is an array of structure pointers, so I will update the declaration of `resolvers` a few lines up.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -137,33 +131,52 @@ jerryx_module_add_to_cache (jerry_value_t cache, /**< cache to which to add the
+  bool return_value = false;
+
+  jerry_size_t c_module_name_size = jerry_get_utf8_string_size (name);
+  jerry_char_t c_module_name[c_module_name_size];
```
While this allocation is simple in source code level, I always has doubts about its efficiency. You need to move a lot of data on the stack.

gabrielschulhof replied:
So, should I use `jmem_heap_alloc_block_null_on_error()` instead?

zherczeg replied:
I am just asking what is its cost. We could introduce convenience stuff for them (I saw you already created some macros for it).

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -180,29 +193,60 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
+  jerry_value_t ret = 0;
+  const jerryx_module_resolver_t *resolver_p;
+  jerry_value_t instances;
+  jerry_value_t canonical_names_static[5];
```
Lets use a define for this.

gabrielschulhof replied:
What do you mean "use a define"?

zherczeg replied:
`#define CANONICAL_NAMES_LOCAL_SIZE 4` or something. I think 4 is enough (probably 2 is the most frequent in practice).

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -180,29 +193,60 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
+
+  for (index = 0; index < count; index++)
+  {
+    canonical_names[index] = 0;
```
0? Not undefined?

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -180,29 +193,60 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
+    resolver_p = resolvers_p[index];
+    canonical_names[index] =
+    resolver_p->get_canonical_name == NULL ? jerry_acquire_value (name) :
+    resolver_p->get_canonical_name (name);
```
Interesting alignment.

gabrielschulhof replied:
If I move the else under `jerry_acquire_value ()` I get this from check-vera.sh:
```
./jerry-ext/module/module.c:235: error: Indentation: 45 -> 4.
```
So, that's why it's like this.

zherczeg replied:
Use round brackets.

martijnthe commented on jerry-ext/module/module.c:
```diff
@@ -137,33 +131,52 @@ jerryx_module_add_to_cache (jerry_value_t cache, /**< cache to which to add the
+jerryx_module_resolver_t jerryx_module_native_resolver =
+{
+  NULL,
+  jerryx_resolve_native_module
```
```
.get_canonical_name = NULL,
.resolve = jerryx_resolve_native_module,
```

martijnthe commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -22,6 +22,24 @@
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name); /**< module name to canonicalize */
+  bool (*resolve) (const jerry_value_t canonical_name, /**< requested module's canonical name */
+                   jerry_value_t *result); /**< resulting module */
```
What's your rationale for passing around the name strings as `jerry_value_t`s?
I think most module resolvers will convert it right back to a C string.
Is it to avoid having to deal with memory management of the C strings?

gabrielschulhof replied:
Yeah, and I was actually thinking that, for path resolution, you might not even convert the whole value to a C string, but you would rather use fixed size chunks to gradually retrieve the string and maybe assemble the canonical name in a similarly gradual fashion. Then, unless you had a directory with a really, really, really long name, you wouldn't really be allocating all that much.

gabrielschulhof replied:
Call it streamed path resolution.

gabrielschulhof replied:
Another very important reason: the module cache is stored as a JavaScript object, so I have to convert a C string back to a `jerry_value_t` in order to check if that JavaScript object has a module at the module name's key.

martijnthe replied:
Ah I see, yeah that makes sense.

gabrielschulhof commented:
@martijnthe @LaszloLango @zherczeg I have updated the patch to reflect the changes to the moduling system.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -92,20 +92,27 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
  * loaded if it is found.
  */
-jerry_value_t jerryx_module_resolve (const jerry_char_t *name, const jerryx_module_resolver_t *resolvers, size_t count);
+jerry_value_t jerryx_module_resolve (const jerry_value_t name, /**< module's name */
```
Ditto

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -92,20 +92,27 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name);  /**< module name to canonicalize */
+  bool (*resolve) (const jerry_value_t canonical_name,  /**< requested module's canonical name */
+                   jerry_value_t *result); /**< resulting module */
```
We do not use comments in header files.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -167,26 +167,38 @@ static const jerry_char_t *on_resolve_absent = (jerry_char_t *) "Module on_resol
 {
   const jerryx_native_module_t *module_p = NULL;
 
+  jerry_size_t name_size = jerry_get_utf8_string_size (canonical_name);
```
Shall we call tostring on the name? Or check if it is a string.

gabrielschulhof replied:
We shouldn't need to. This function should not be called directly, even though a pointer to it is available. It should only be called by `jerryx_module_resolve ()` and it should only ever receive a `jerry_value_t` stored in the array of canonical module names created by `jerryx_module_resolve ()`. In fact, this function is not even documented. Only the structure `jerryx_module_native_resolver` is documented, and only the availability of it as a global data symbol is documented, not its contents.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -202,29 +214,51 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
-jerryx_module_resolve (const jerry_char_t *name, /**< name of the module to load */
-                       const jerryx_module_resolver_t *resolvers_p, /**< list of resolvers */
+jerryx_module_resolve (const jerry_value_t name, /**< name of the module to load */
+                       const jerryx_module_resolver_t **resolvers_p, /**< list of resolvers */
```
`**` ? A pointer to a list of pointers which points to the resolvers? Why not an array of resolvers?

gabrielschulhof replied:
Because the resolvers are stored in-contiguously, and we do not want to make copies of them, right?

zherczeg replied:
Are they used multiple times? Or why they are scattered?

I think one application use one resolver list, but this is not much of a code size increase, so lleave it be.

gabrielschulhof replied:
Resolvers may be declared global in separate .c files (jerryx_module_native_resolver` certainly is), and they are collected into an array only once - the place that calls `jerryx_module_resolve ()`.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -202,29 +214,51 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
+    resolver_p = resolvers_p[index];
+    canonical_names[index] = ((resolver_p && resolver_p->get_canonical_name != NULL) ?
+      resolver_p->get_canonical_name (name):
+      jerry_acquire_value (name));
```
This is not how we align ?:

gabrielschulhof replied:
It's pretty tough not breaking 120 columns with this line. I'm still trying to figure out how to align it.

gabrielschulhof replied:
I renamed `resolver_p` to `res_p` then it fits.

zherczeg replied:
I would move the condition into a bool value rather than shoretning the name.

zherczeg commented on tests/unit-ext/module/jerry-module-test.c:
```diff
@@ -61,41 +61,61 @@ const char eval_string5[] =
 static bool
-resolve_differently_handled_module (const jerry_char_t *name,
+resolve_differently_handled_module (const jerry_value_t name,
                                                 jerry_value_t *result)
```
Comment is missing, spaces are wrong.

gabrielschulhof commented:
@zherczeg I have addressed your review comments.

gabrielschulhof commented:
@zherczeg instead of a boolean I went straight for storing the function pointers in variables.

gabrielschulhof commented:
... aaand I forgot to key the cache on the canonical name. Fixed now.

gabrielschulhof commented:
@zherczeg could you or somebody with Travis access please restart the lint job? It seems to have been killed because it had stalled.

gabrielschulhof commented:
Thanks!

LaszloLango commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -29,9 +29,9 @@ typedef jerry_value_t (*jerryx_native_module_on_resolve_t) (void);
-  struct jerryx_native_module_t *next_p; /**< pointer to next module in the list */
+  const jerry_char_t *name_p;
+  const jerryx_native_module_on_resolve_t on_resolve;
+  struct jerryx_native_module_t *next_p;
```
Why did you remove the comments from here?

gabrielschulhof replied:
@zherczeg [said](https://github.com/jerryscript-project/jerryscript/pull/2013#discussion_r140710783) we didn't add such `/**<` comments to the header files. Or did I misinterpret?

LaszloLango replied:
Yes, I think there was a misunderstanding here. We do not add comments to declarations, but only to definitions. The type definitions are usually in the header files, so we must put the comments there. Please don't remove these comments.

zherczeg replied:
Sorry, I mean only for function arguments, not for types.

LaszloLango commented on jerry-ext/module/module.c:
```diff
@@ -14,11 +14,13 @@
  */
 
 #include <string.h>
+#include "jmem.h"
```
I think we should not use internal headers in `jerry-ext`. I don't see what are you using from it. Please remove it if possible.

LaszloLango commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -92,20 +92,27 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
-bool jerryx_module_native_resolver (const jerry_char_t *name, jerry_value_t *result);
+typedef struct
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name);
```
Missing comments

gabrielschulhof replied:
I need to know for sure. Do we, or do we not add `/**<` comments after structure members and function arguments? How do we document function pointer structure members, since the formal parameters of the function signature also needs to be documented?

gabrielschulhof commented:
@LaszloLango fixed up the comments and added some `_p`s.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -92,20 +92,41 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
+ */
+typedef jerry_value_t (*jerryx_module_get_canonical_name_t) (const jerry_value_t name); /**< The name for which to
+                                                                                         * compute the canonical
+                                                                                         * name */
```
Align the start of the comment lines to the same column.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -92,20 +92,41 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
-bool jerryx_module_native_resolver (const jerry_char_t *name, jerry_value_t *result);
+typedef bool (*jerryx_module_resolve_t) (const jerry_value_t canonical_name, /**< The module's canonical name */
+                                         jerry_value_t *result); /**< The resulting module, if the function returns
+                                                                  * true */
```
Ditto. And similar to below.

zherczeg commented on tests/unit-ext/test-ext-module-canonical.c:
```diff
@@ -0,0 +1,93 @@
+
+  return (!strcmp ((char *) name_string, "batman") ? jerry_acquire_value (name):
+         (!strcmp ((char *) name_string, "bruce-wayne") ? jerry_create_string ((jerry_char_t *) "batman"):
+         jerry_create_undefined ()));
```
Fix the style of these lines.

zherczeg commented on tests/unit-ext/test-ext-module-canonical.c:
```diff
@@ -0,0 +1,93 @@
+  jerry_init (JERRY_INIT_EMPTY);
+
+  jerry_value_t batman = jerry_create_string ((jerry_char_t *) "batman");
+  jerry_value_t bruce_wayne = jerry_create_string ((jerry_char_t *) "bruce-wayne");
```
I hope these are not copyrighted.

gabrielschulhof replied:
Good point. I'll use [Liberator](http://pdsh.wikia.com/wiki/Liberator), which is public domain.

zherczeg replied:
They are still, however, "use at your own risk" characters. - this doesn't sound entirely public

gabrielschulhof replied:
*sigh* Alice and Bob it is.

gabrielschulhof commented:
@zherczeg @LaszloLango please have another look. I updated the test to use `#define`-ed strings to store the alias and the actual name at the top, and I replaced the ternary composition with `if`/`else if`/`else`. I also changed the resolver initializer to use the `{ .member_name = value }` notation.

LaszloLango commented:
still LGTM

zherczeg commented:
Still ok.
